Correction of Location of Zone Equipment Get Routine

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #10459 
 - This pull request contains a fix to an EnergyPlus crash that a user ran into in an input file that contains VRF equipment and no internal gains.  This combination caused the crash though it is possible that other combinations might also cause the same thing to happen.  In a nutshell, the get routine for zone equipment wasn't being triggered for this combination and patches have been installed in the code over the years to cover other cases.  This fix removes all of the patches and moves the Zone Equipment get call up higher to avoid this happening in any other model.  With this fix, the Zone Equipment get happens right after zones and geometry is set-up, essentially guaranteeing that other things in HVAC that need zone equipment to be gotten already will have what they need to function properly.  After the fix, there were no differences in the output files aside from a few BND and RDD files which showed some slight reordering of certain lines.  All ESO files were identical after the fix.  A unit test verifies that the get routine is only triggered once and thus prevents reallocation of arrays that cause a crash (the crash in the unit test was confirmed by resetting the variable that prevents subsequent calls to the get routine--though obviously this is not part of the pull request).  No documentation changes were needed for this fix.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Fix of problem

When there are no internal gains are present and VRF is used, E+ crashes because some structures are not allocated before it gets to the VRF code.  Code was added to resolve this issue by getting the zone equipment data.  Other examples of this strategy already exist in the code.

RKStrand committed:
Moved get routine to HVAC Manager

Removed the get zone equipment call out of HVACVariableRefrigerantFlow and put a new call in to the zone equipment get routine.  Next step will get to remove all other get zone equipment calls and see what happens.  Doing this in stages to hopefully avoid other issues.

RKStrand committed:
Moving GetZoneEquipment on up

The call to GetZoneEquipment has been moved up to the HeatBalanceManager.cc and called from GetBuildingData right after the call to GetZoneData.  All other calls in the code to GetZoneEquipment have been removed.  Various unit tests were altered to adjust for this new reality.  Now, only one call is ever made and other models do not have to patch in a call.

RKStrand committed:
Merge branch 'develop' into 10459ZoneEquipNeedsInitInVRF

RKStrand committed:
Corrected the order of calls

The order of subroutine calls was slightly off, causing one of the test suite files to crash.  This fixes the issue.  Hopefully it won't introduce any other problems...

RKStrand committed:
Addition of unit test

Added a unit test.  It runs GetZoneEquipment and properly resets the get flag.  Then the second call to GetZoneEquipment, it doesn't do anything which is correct.  If it goes into GetZoneEquipmentData which is called by GetZoneEquipment, the program crashes which would be correct.  So, this tests that only one time into GetZoneEquipmentData is allowed.

rraustad commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -152,12 +152,6 @@ void ManageZoneEquipment(EnergyPlusData &state,
-        state.dataZoneEquipmentManager->GetZoneEquipmentInputFlag = false;
-        state.dataZoneEquip->ZoneEquipInputsFilled = true;
-    }
-
```
Move from ManageZoneEquipment to HeatBalanceManager::GetBuildingData. That is certainly earlier in the simulation. And then you removed all the other calls to GetZoneEquipmentData. You Sir, have nerves of steel!

Myoldmopar replied:
That's what I was thinking!

rraustad commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -1972,6 +1973,8 @@ namespace HeatBalanceManager {
 
         SurfaceGeometry::SetupZoneGeometry(state, ErrorsFound);
+
+        ZoneEquipmentManager::GetZoneEquipment(state);
```
Now called from GetBuildingData. And GetBuildingData is called early in the simulation inside GetHeatBalanceInput which calls several setup functions.

    void GetHeatBalanceInput(EnergyPlusData &state)
    {
        GetProjectControlData(state, ErrorsFound);

        GetSiteAtmosphereData(state, ErrorsFound);

        GetWindowGlassSpectralData(state, ErrorsFound);

        Material::GetMaterialData(state, ErrorsFound); // Read materials from input file/transfer from legacy data structure

        GetFrameAndDividerData(state);

        GetConstructData(state, ErrorsFound); // Read constructs from input file/transfer from legacy data structure

        GetBuildingData(state, ErrorsFound); // Read building data from input file

        DataSurfaces::GetVariableAbsorptanceSurfaceList(state);

        GetIncidentSolarMultiplier(state, ErrorsFound);

        // Added SV 6/26/2013 to load scheduled surface gains
        GetScheduledSurfaceGains(state, ErrorsFound);

        if (state.dataSurface->UseRepresentativeSurfaceCalculations) {

rraustad reviewed (approved):
This looks really good. Move basic object calls to earlier in the simulation to avoid duplicate code.

rraustad commented:
So now I see a call to GetZoneEquipmentData in only 2 places. The place where you moved it to here, and in GetControlledZoneIndex which is called from, wait for it ..., GetInternalHeatGainsInput and also GetWindowGapAirflowControlData, both of which get a zone index. Being unsure of calling order I am not sure if it should be deleted there as well (I'm thinking it's likely, but definitely trace the call order before doing anything). Given that GetControlledZoneIndex is a mining function there is nothing wrong with that call residing there as well.

    int GetControlledZoneIndex(EnergyPlusData &state, std::string const &ZoneName) // Zone name to match into Controlled Zone structure
    {
        if (!state.dataZoneEquip->ZoneEquipInputsFilled) {
            GetZoneEquipmentData(state);
            state.dataZoneEquip->ZoneEquipInputsFilled = true;
        }

        return Util::FindItemInList(ZoneName, state.dataZoneEquip->ZoneEquipConfig, &EquipConfiguration::ZoneName);
    }

rraustad commented:
@Myoldmopar this is ready to merge.

RKStrand commented:
> So now I see a call to GetZoneEquipmentData in only 2 places. The place where you moved it to here, and in GetControlledZoneIndex which is called from, wait for it ..., GetInternalHeatGainsInput and also GetWindowGapAirflowControlData, both of which get a zone index. Being unsure of calling order I am not sure if it should be deleted there as well (I'm thinking it's likely, but definitely trace the call order before doing anything). Given that GetControlledZoneIndex is a mining function there is nothing wrong with that call residing there as well.
> 
> ```
> int GetControlledZoneIndex(EnergyPlusData &state, std::string const &ZoneName) // Zone name to match into Controlled Zone structure
> {
>     if (!state.dataZoneEquip->ZoneEquipInputsFilled) {
>         GetZoneEquipmentData(state);
>         state.dataZoneEquip->ZoneEquipInputsFilled = true;
>     }
> 
>     return Util::FindItemInList(ZoneName, state.dataZoneEquip->ZoneEquipConfig, &EquipConfiguration::ZoneName);
> }
> ```

Well, that is a surprise/disappointment because I thought I got them all.  I looked into those and here is what I can say:

- The call in GetInternalGainsInput happens after the new placement of GetZoneEquipment which calls GetZoneEquipmentData.  So, I think we can get rid of this call now.
- The call from GetWindowGapAirflowControlData happens from SetupZoneGeometry.  Now in this case, I had to move the call to GetZoneEquipment until right after SetupZoneGeometry because there was one case in the input that crashed when I had the order switched and GetZoneEquipment happened right before SetupZoneGeometry.  I would need to test some things to see if we can get this to work.  I get it that the air gap stuff will need the zone equipment stuff set.

So, I will do a bit of additional testing this morning and see what happens.  While it makes more sense to call GetZoneEquipment from where I put it, I guess if I can move it to SetupZoneGeometry and find the one place that will eliminate all other calls to GetZoneEquipment that might be preferable.

RKStrand committed:
Now all duplicates removed

Missed two duplicate calls.  This should now get all zone equipment gets in one place.

RKStrand commented:
@rraustad Ok...I just made a commit that eliminates the remaining duplicates.  Hoping this comes back all green...

rraustad commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -108,6 +108,7 @@
 #include <EnergyPlus/WindowComplexManager.hh>
 #include <EnergyPlus/WindowEquivalentLayer.hh>
 #include <EnergyPlus/WindowManager.hh>
+#include <EnergyPlus/ZoneEquipmentManager.hh>
```
And this include is no longer needed.

RKStrand replied:
Oops...I knew I was going to forget something small like that.  Taken care of in the new commit just made.  Thanks for spotting that.

RKStrand committed:
Clean up on aisle HBMan

Forgot to remove an include.

rraustad commented on tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:
```diff
@@ -5268,3 +5268,63 @@ TEST_F(EnergyPlusFixture, SpaceHVACMixerTest)
+    AllocateHeatBalArrays(*state);
+    GetZoneEquipment(*state);
+    EXPECT_FALSE(ErrorsFound);
+    EXPECT_FALSE(state->dataZoneEquipmentManager->GetZoneEquipmentInputFlag);
```
1) is this unit test still applicable to the latest code change?, and 2) I think it would be better to verify what arrays are allocated/filled with the data expected to be available to the simulation at the point where everything is now read in.

e.g.:

    EXPECT_GT(int(state.dataZoneEquip->ZoneEquipConfig.size()), 0); 
    EXPECT_EQ(state.dataZoneEquip->ZoneEquipConfig(1).ZoneNode, 1);

same for equipment list and whatever else is now available. This would still be a weak unit test but at least it would lock in what the current early call does.

RKStrand replied:
I'll look at this some more.  My thought is that it is still valid as there are other places in the unit tests where it calls GetZoneEquipment directly rather than via the new location in SetupZoneGeometry.  I think it's still ok to do that.  I will see if I can expand the unit test a bit.

RKStrand replied:
But on the positive side...things seem to still be coming back green so far...

rraustad replied:
Well, it was just a suggestion. What you did with getting the zone inputs early is more important to me than this unit test. Not all code changes require a unit test.

RKStrand replied:
Ok...added some stuff here as requested.  It's still pretty basic, but this is better at verifying that it really isn't going through the actual code of GetZoneEquipment a second time.  Better?

RKStrand committed:
Improvement to the unit test

Added some more stuff here to make the unit test a little less weak.

rraustad commented on tst/EnergyPlus/unit/ZoneEquipmentManager.unit.cc:
```diff
@@ -5268,3 +5268,74 @@ TEST_F(EnergyPlusFixture, SpaceHVACMixerTest)
+    GetZoneEquipment(*state);
+    EXPECT_FALSE(state->dataZoneEquipmentManager->GetZoneEquipmentInputFlag);
+    EXPECT_EQ(state->dataZoneEquipmentManager->NumOfTimeStepInDay, 24);
+}
```
Yes, this is better.

rraustad commented:
@Myoldmopar this is ready to merge.

rraustad commented:
Verified that the defect file tries to access an unallocated array using develop.
Verified the defect file runs successfully in this branch (debug and release).
Verified unit tests run successfully.

Myoldmopar commented:
OK, this one is up next.  Testing locally now.

Myoldmopar commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -314,6 +315,8 @@ namespace SurfaceGeometry {
             return;
         }
 
+        ZoneEquipmentManager::GetZoneEquipment(state); // Necessary to get this before window air gap code
```
I haven't looked yet, but I assume this is causing CI diffs just due to reordering, right?

Myoldmopar commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -183,19 +177,21 @@ void GetZoneEquipment(EnergyPlusData &state)
+        }
 
-    state.dataZoneEquipmentManager->PrioritySimOrder.allocate(MaxNumOfEquipTypes);
+        state.dataZoneEquipmentManager->PrioritySimOrder.allocate(MaxNumOfEquipTypes);
```
That's it?  That's the whole change here?  OK, love the small amount of code to review!

Myoldmopar reviewed (approved):
Seems fine, and builds and runs locally fine.  Checking CI results and then staging for merge.

Myoldmopar commented:
Excellent, just BND and EDD diffs.  I'll take it!  I think this one is ready to go now.  Merging it in, thanks @RKStrand and @rraustad !

RKStrand commented:
Thanks, Edwin! This was a very satisfying defect for a variety of reasons, but as always I’m glad to have it merged.

Also, just a heads up that today is the last Wednesday of early school dismissal which falls smack dab in the middle of today’s technicalities call. So, I won’t be present today. However, since I’m not doing development this FY and I don’t have any bugs that I’ve started and haven’t been merged at the moment (that’ll change tomorrow hopefully), there’s really nothing that I have to present. Anyway, happy meeting and I’ll be back on the call next Wednesday.

Rick

Sent from my iPhone

On May 1, 2024, at 10:47 AM, Edwin Lee ***@***.***> wrote:

﻿

Merged #10482<https://urldefense.com/v3/__https://github.com/NREL/EnergyPlus/pull/10482__;!!DZ3fjg!6dNfh8iOS0xUdZipt51Dzu16p5glwUwsY4L4z9b92c7jST7sKnCal-UnU4JqCRDmp5Wo79Ga55z_Cyg8BChJFRHFPnU$> into develop.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/NREL/EnergyPlus/pull/10482*event-12671949132__;Iw!!DZ3fjg!6dNfh8iOS0xUdZipt51Dzu16p5glwUwsY4L4z9b92c7jST7sKnCal-UnU4JqCRDmp5Wo79Ga55z_Cyg8BChJzduk1hg$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ABMJHPJZBXTP4F7I7PS7PHDZAEFBBAVCNFSM6AAAAABGZV52WKVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJSGY3TCOJUHEYTGMQ__;!!DZ3fjg!6dNfh8iOS0xUdZipt51Dzu16p5glwUwsY4L4z9b92c7jST7sKnCal-UnU4JqCRDmp5Wo79Ga55z_Cyg8BChJRCU-LyI$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
